In [1]:
import sys
print(sys.executable)

D:\pytorch_project\pytorch_env\Scripts\python.exe


In [2]:
import pandas as pd

In [3]:
amazon_purchases_data = pd.read_csv(r"C:\Users\Atharva\OneDrive\Desktop\BDA CP\archive (1)\amazon-purchases.csv")
servey_data= pd.read_csv(r"C:\Users\Atharva\OneDrive\Desktop\BDA CP\archive (1)\survey.csv")

In [4]:
amazon_purchases_data.head()

,Order Date,Purchase Price Per Unit,Quantity,Shipping Address State,Title,ASIN/ISBN (Product Code),Category,Survey ResponseID
0,2018-12-04,7.98,1.0,NJ,SanDisk Ultra 16GB Class 10 SDHC UHS-I Memory ...,B0143RTB1E,FLASH_MEMORY,R_01vNIayewjIIKMF
1,2018-12-22,13.99,1.0,NJ,Betron BS10 Earphones Wired Headphones in Ear ...,B01MA1MJ6H,HEADPHONES,R_01vNIayewjIIKMF
2,2018-12-24,8.99,1.0,NJ,NaN,B078JZTFN3,NaN,R_01vNIayewjIIKMF
3,2018-12-25,10.45,1.0,NJ,Perfecto Stainless Steel Shaving Bowl. Durable...,B06XWF9HML,DISHWARE_BOWL,R_01vNIayewjIIKMF
4,2018-12-25,10.00,1.0,NJ,Proraso Shaving Cream for Men,B00837ZOI0,SHAVING_AGENT,R_01vNIayewjIIKMF


In [5]:
servey_data.head()

,Survey ResponseID,Q-demos-age,Q-demos-hispanic,Q-demos-race,Q-demos-education,Q-demos-income,Q-demos-gender,Q-sexual-orientation,Q-demos-state,Q-amazon-use-howmany,...,Q-substance-use-marijuana,Q-substance-use-alcohol,Q-personal-diabetes,Q-personal-wheelchair,Q-life-changes,Q-sell-YOUR-data,Q-sell-consumer-data,Q-small-biz-use,Q-census-use,Q-research-society
0,R_1ou69fj4DQGsVcp,35 - 44 years,No,Black or African American,High school diploma or GED,"$25,000 - $49,999",Female,heterosexual (straight),Iowa,2,...,No,Yes,No,No,Lost a job,No,No,No,No,No
1,R_2UbJL30HRjK1sdD,45 - 54 years,No,White or Caucasian,High school diploma or GED,"$100,000 - $149,999",Male,heterosexual (straight),Ohio,2,...,No,No,No,No,NaN,No,No,No,No,Yes
2,R_UPXamGKtmf4RVIZ,25 - 34 years,No,White or Caucasian,High school diploma or GED,"$25,000 - $49,999",Male,heterosexual (straight),Arkansas,1 (just me!),...,No,No,Yes,No,NaN,No,No,No,No,Yes
3,R_2dYk5auG9Fv5Qve,35 - 44 years,Yes,White or Caucasian,"Graduate or professional degree (MA, MS, MBA, ...","$50,000 - $74,999",Male,heterosexual (straight),Tennessee,1 (just me!),...,No,No,No,No,NaN,No,No,No,No,No
4,R_2aP0GyIR66gSTiR,25 - 34 years,No,White or Caucasian,High school diploma or GED,"$50,000 - $74,999",Male,heterosexual (straight),Virginia,2,...,No,Yes,No,No,NaN,No,Yes if consumers get part of the profit,I don't know,No,No


In [6]:
amazon_purchases_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1850717 entries, 0 to 1850716
Data columns (total 8 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   Order Date                object 
 1   Purchase Price Per Unit   float64
 2   Quantity                  float64
 3   Shipping Address State    object 
 4   Title                     object 
 5   ASIN/ISBN (Product Code)  object 
 6   Category                  object 
 7   Survey ResponseID         object 
dtypes: float64(2), object(6)
memory usage: 113.0+ MB


The Dtype of Order data is object here which should be converted to data_time format

In [7]:
amazon_purchases_data['Order Date']= pd.to_datetime(amazon_purchases_data['Order Date'])
amazon_purchases_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1850717 entries, 0 to 1850716
Data columns (total 8 columns):
 #   Column                    Dtype         
---  ------                    -----         
 0   Order Date                datetime64[ns]
 1   Purchase Price Per Unit   float64       
 2   Quantity                  float64       
 3   Shipping Address State    object        
 4   Title                     object        
 5   ASIN/ISBN (Product Code)  object        
 6   Category                  object        
 7   Survey ResponseID         object        
dtypes: datetime64[ns](1), float64(2), object(5)
memory usage: 113.0+ MB


Alright we changed it to datetime dtype

In [8]:
amazon_purchases_data.describe().T

,count,mean,min,25%,50%,75%,max,std
Order Date,1850717,2020-12-03 02:27:02.550828032,2018-01-01 00:00:00,2019-12-06 00:00:00,2021-01-20 00:00:00,2022-01-17 00:00:00,2024-08-15 00:00:00,NaN
Purchase Price Per Unit,1850717.0,22.659063,0.01,8.47,13.99,23.75,6398.95,45.999671
Quantity,1850717.0,1.088636,1.0,1.0,1.0,1.0,339.0,0.779534


In [9]:
servey_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5027 entries, 0 to 5026
Data columns (total 23 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Survey ResponseID           5027 non-null   object
 1   Q-demos-age                 5027 non-null   object
 2   Q-demos-hispanic            5027 non-null   object
 3   Q-demos-race                5027 non-null   object
 4   Q-demos-education           5027 non-null   object
 5   Q-demos-income              5027 non-null   object
 6   Q-demos-gender              5027 non-null   object
 7   Q-sexual-orientation        5027 non-null   object
 8   Q-demos-state               5027 non-null   object
 9   Q-amazon-use-howmany        5027 non-null   object
 10  Q-amazon-use-hh-size        5027 non-null   object
 11  Q-amazon-use-how-oft        5027 non-null   object
 12  Q-substance-use-cigarettes  5027 non-null   object
 13  Q-substance-use-marijuana   5027 non-null   obje

In [10]:
servey_data.describe().T

,count,unique,top,freq
Survey ResponseID,5027,5027,R_273TMGuLM3jXisi,1
Q-demos-age,5027,6,25 - 34 years,1813
Q-demos-hispanic,5027,2,No,4478
Q-demos-race,5027,25,White or Caucasian,3886
Q-demos-education,5027,5,Bachelor's degree,2219
Q-demos-income,5027,7,"$25,000 - $49,999",1189
Q-demos-gender,5027,4,Female,2589
Q-sexual-orientation,5027,3,heterosexual (straight),3858
Q-demos-state,5027,52,California,505
Q-amazon-use-howmany,5027,4,1 (just me!),3546


In [11]:
amazon_purchases_data.isna().sum()

Order Date                      0
Purchase Price Per Unit         0
Quantity                        0
Shipping Address State      87812
Title                       89740
ASIN/ISBN (Product Code)      973
Category                    89458
Survey ResponseID               0
dtype: int64

There are 4 columns with null values here those being
- Shipping_address
- Title
- ASIN/ISBN (Product Code)
- Category

Now I have to see if the address issue is from the same user because it might be that same user order multiple times. Same goes with all the other products I just have to confirm these things. 

In [12]:
multi_null_users=(
    amazon_purchases_data[amazon_purchases_data["Shipping Address State"].isna()].groupby("Survey ResponseID").size()
)
multi_null_users

Survey ResponseID
R_01vNIayewjIIKMF    10
R_037XK72IZBJyF69     9
R_038ZU6kfQ5f89fH    35
R_03aEbghUILs9NxD     9
R_06RZP9pS7kONINr     4
                     ..
R_zbZLoKX4IzA4oql     6
R_zczhhGpVCmLwad3     2
R_zd4E1BgAdaM2761     3
R_zfnsck2ZatnGtY5    16
R_zfqnsBzlOAKibzb     1
Length: 3730, dtype: int64

In [13]:
multi_null_users.shape[0]

3730

There are 3730 users with no shipping address data and not 87k+ now lets see if there is any record of any of their address in rest of the data

user_has_address = (
    amazon_purchases_data
    .groupby("Survey ResponseID")["Shipping Address State"]
    .apply(lambda x: x.notna().any())
)

user_has_address

We basically used this to see if there are any users which have provided their address atleast once if they have it says true if not it says false then next we will see out of those 3730 users how many have not provided address even once

In [14]:
user_has_address = (
    amazon_purchases_data
    .groupby("Survey ResponseID")["Shipping Address State"]
    .apply(lambda x: x.notna().any())
)

In [15]:
int((~user_has_address).sum())

62

So out of 3730 users only 62 have not provided any address even once. Lets look at how many entires do these 62 users have

In [16]:
# Get those users (False ones)
users_without_address = user_has_address[~user_has_address].index
users_without_address[:10]

Index(['R_10GfwNctrOL9e2t', 'R_12G93XcmsS5F4Ea', 'R_1FPtnyDUi83F045',
       'R_1JDPvSxDgeKBb7U', 'R_1KlMYPvPKoudQO5', 'R_1LbUTgaWmOcYHoQ',
       'R_1d1fnT4sjZABBwe', 'R_1hT2aJkf3L95zWl', 'R_1jYOsGh5TQW4Hsg',
       'R_1kNwFGBAUSrxQP0'],
      dtype='object', name='Survey ResponseID')

In [17]:
# Filter original dataset
entries_from_these_users = amazon_purchases_data[
    amazon_purchases_data["Survey ResponseID"].isin(users_without_address)
]
entries_from_these_users[:3]

,Order Date,Purchase Price Per Unit,Quantity,Shipping Address State,Title,ASIN/ISBN (Product Code),Category,Survey ResponseID
20927,2018-01-11,8.88,1.0,NaN,"Mrs. Meyer's Hand Soap, Made with Essential Oi...",B00NU9MJYS,SKIN_CLEANING_AGENT,R_10GfwNctrOL9e2t
20928,2018-01-11,7.56,1.0,NaN,"Tom's of Maine Whole Care Fluoride Toothpaste,...",B00JUJ140M,TOOTH_CLEANING_AGENT,R_10GfwNctrOL9e2t
20929,2018-01-11,10.60,1.0,NaN,CLIF BARS - Energy Bars - Coconut Chocolate Ch...,B0050MM18M,SNACK_FOOD_BAR,R_10GfwNctrOL9e2t


In [18]:
# Count rows
entries_from_these_users.shape[0]

24182

tose 62 users have 24182 orders which is quite a bit lets see how we can handle them later but first the other users lets fill the values for the once who have their address written at least once

In [19]:
# Find users who have BOTH null and non-null
user_status = amazon_purchases_data.groupby("Survey ResponseID")["Shipping Address State"].agg(
    has_null=lambda x: x.isna().any(),
    has_value=lambda x: x.notna().any()
)
user_status[:5]

,has_null,has_value
Survey ResponseID,,
R_01vNIayewjIIKMF,True,True
R_037XK72IZBJyF69,True,True
R_038ZU6kfQ5f89fH,True,True
R_03aEbghUILs9NxD,True,True
R_06RZP9pS7kONINr,True,True


In [20]:
# Users that can be filled

fillable_users = user_status[
    (user_status["has_null"]) & (user_status["has_value"])
].index
fillable_users

Index(['R_01vNIayewjIIKMF', 'R_037XK72IZBJyF69', 'R_038ZU6kfQ5f89fH',
       'R_03aEbghUILs9NxD', 'R_06RZP9pS7kONINr', 'R_06d9ULxrBmkwSTn',
       'R_07oHvj3bLVVRCRb', 'R_085qq7w0pkhowox', 'R_08uYA7fb4unHGkF',
       'R_0BCm9hRP9iTI081',
       ...
       'R_zTf2i6s3E2OQZQR', 'R_zTi3j2QuqAzr7NL', 'R_zYlhkltqLl0WShX',
       'R_zZSx9DRg5AejOrD', 'R_zZsd6BQmBdvVG0h', 'R_zbZLoKX4IzA4oql',
       'R_zczhhGpVCmLwad3', 'R_zd4E1BgAdaM2761', 'R_zfnsck2ZatnGtY5',
       'R_zfqnsBzlOAKibzb'],
      dtype='object', name='Survey ResponseID', length=3668)

In [21]:
# Count number of such users

num_fillable_users = len(fillable_users)
num_fillable_users

3668

In [22]:
# Count number of entries (rows) that will be filled

entries_to_fill = amazon_purchases_data[
    (amazon_purchases_data["Survey ResponseID"].isin(fillable_users)) &
    (amazon_purchases_data["Shipping Address State"].isna())
]
entries_to_fill[:3]

,Order Date,Purchase Price Per Unit,Quantity,Shipping Address State,Title,ASIN/ISBN (Product Code),Category,Survey ResponseID
32,2019-10-05,25.0,1.0,NaN,Best Buy Gift Cards - Email Delivery,B07RTN4C77,GIFT_CARD,R_01vNIayewjIIKMF
33,2019-10-06,50.0,1.0,NaN,Best Buy Gift Cards - Email Delivery,B07RTN4C77,GIFT_CARD,R_01vNIayewjIIKMF
34,2019-10-06,50.0,1.0,NaN,Best Buy Gift Cards - Email Delivery,B07RTN4C77,GIFT_CARD,R_01vNIayewjIIKMF


In [23]:
num_entries_to_fill = len(entries_to_fill)
num_entries_to_fill

63630

We will be recovering 63630 entries which is a lot tbh but I do have one concern what if they changed address or when they ordered the items they were at different address. Now we can just speculate things. I am assuming they had the same address as I do not think there is a way to find that out and we cannot be losing this huge data for that alone

I later realised that we did have that information. I will later chack if it matters or not I dont think it should but lets see

In [24]:
num_entries_to_fill / len(amazon_purchases_data)

0.034381269529593125

We basically saved aprrox 3.4% of our data with this and some of them might be important users that buy more things which would be good for building recommendation engine

In [25]:
#Fill using group-level information
filled_series = (
    amazon_purchases_data
    .groupby("Survey ResponseID")["Shipping Address State"]
    .transform(lambda x: x.ffill().bfill())
)

C:\Users\Atharva\AppData\Local\Temp\ipykernel_3308\2279527257.py:5: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .transform(lambda x: x.ffill().bfill())


In [26]:
amazon_purchases_data.loc[
    amazon_purchases_data["Shipping Address State"].isna(),
    "Shipping Address State"
] = filled_series

In [27]:
# Remaining null values
amazon_purchases_data["Shipping Address State"].isna().sum()

np.int64(24182)

We went from 87812 null values to 24182 for this section we will see what to do with these but now lets focus on other values now

Lets see if every title has unique product id, if they do we might be able to recover some of that data too

In [28]:
# Count unique codes per Title
title_code_counts = (
    amazon_purchases_data
    .groupby("Title")["ASIN/ISBN (Product Code)"]
    .nunique()
)
title_code_counts[:10]

Title
 Hypergiant 12W Waterproof Plug in UV LED Spotlight ,Outdoor Black Lights,120V AC,3.3ft wire,Stake landscape Light for Outdoor Decoration Yard Lawn Patio Garden Path Tree(2-PACK)    1
Asurion   Lawn & Garden Extended Protection Plan 4 Year                                                                                                                               1
Care Check Lens Wipes, 300 Pre-Moistened Cleaning Wipes for Cameras, Laptops, Cell Phones, Eyeglasses, Other Screens and More                                                         1
Fisher-Price Little People Loops 'n Swoops Amusement Park [Amazon Exclusive]                                                                                                          1
Fruit of the Loom Men's Short Sleeve Crew Tee, Large  - Neon Green                                                                                                                    1
Fruit of the Loom Men's Short Sleeve Crew Tee, X-Large  - Retro Hth Purp  

In [29]:
# Check if all are unique
(title_code_counts > 1).sum()

np.int64(36060)

Not all of them are unique however it is possible that same product with different version might be having slightly different name but same product id. Lets check that

In [30]:
title_code_counts.describe()

count    824722.000000
mean          1.099424
std           0.989648
min           0.000000
25%           1.000000
50%           1.000000
75%           1.000000
max         443.000000
Name: ASIN/ISBN (Product Code), dtype: float64

On average the titles have one product id but small sum has multiple which is increasing the mean to almost 1.1. std = 0.989, min = 0: Some titles have 0 product codes(this is the next thing we will come to after fixing this). 25%, 50%, 75% = 1: For 75% of titles → exactly 1 product code. Most titles are clean and consistent.Variation is quite high relative to the mean This might confirm our suspicion of different versions lets check if we are correct

In [31]:
title_code_counts.sort_values(ascending=False).head(10)

Title
1                                                                                                                      443
Amazon.com eGift Card                                                                                                  265
Amazon.com Gift Card in a Greeting Card (Various Designs)                                                               94
Amazon Essentials Women's Cotton Bikini Brief Underwear (Available in Plus Size), Multipacks                            85
Amazon.com Print at Home Gift Card                                                                                      83
Hanes Men's Sweatshirt, EcoSmart Fleece Hoodie, Cotton-Blend Fleece Hooded Sweatshirt, Plush Fleece Pullover Hoodie     76
Crocs Unisex-Adult Classic Clogs (Best Sellers)                                                                         73
Leggings Depot Women's Printed Solid Activewear Jogger Track Cuff Sweatpants                                            72
Levi's Men

These are not normal “products”, they are:

Generic / template products
- Gift cards
- clothing lines
- bulk items

These naturally have:

- multiple variations
- multiple ASINs

In [32]:
(title_code_counts == 1).sum()

np.int64(788363)

In [33]:
total_titles = title_code_counts.shape[0]
total_titles

824722

In [34]:
788363 / 824722

0.9559136290774346

Over 95% of data hass single product code so we are safe to put same product code and others are outliers. I do not know if I should be removing outliers before or after fillings the null values 

In [35]:
# Build status per Title
title_asin_status_df = (
    amazon_purchases_data
    .groupby("Title")["ASIN/ISBN (Product Code)"]
    .agg(
        title_has_null_asin=lambda x: x.isna().any(),
        title_has_valid_asin=lambda x: x.notna().any()
    )
)

In [36]:
# Classification (new variables)

titles_all_asin_present = title_asin_status_df[
    (~title_asin_status_df["title_has_null_asin"]) &
    (title_asin_status_df["title_has_valid_asin"])
]

titles_all_asin_missing = title_asin_status_df[
    (title_asin_status_df["title_has_null_asin"]) &
    (~title_asin_status_df["title_has_valid_asin"])
]

titles_partial_asin = title_asin_status_df[
    (title_asin_status_df["title_has_null_asin"]) &
    (title_asin_status_df["title_has_valid_asin"])
]

In [37]:
# Counts
num_titles_all_asin_present = len(titles_all_asin_present)
num_titles_all_asin_missing = len(titles_all_asin_missing)
num_titles_partial_asin = len(titles_partial_asin)


In [38]:
num_titles_all_asin_present


824139

In [39]:
num_titles_all_asin_missing


299

In [40]:
num_titles_partial_asin

284

In [41]:
# ASIN → Title
# Build status per ASIN

asin_title_status_df = (
    amazon_purchases_data
    .groupby("ASIN/ISBN (Product Code)")["Title"]
    .agg(
        asin_has_null_title=lambda x: x.isna().any(),
        asin_has_valid_title=lambda x: x.notna().any()
    )
)

In [42]:
# Classification

asin_all_title_present = asin_title_status_df[
    (~asin_title_status_df["asin_has_null_title"]) &
    (asin_title_status_df["asin_has_valid_title"])
]

asin_all_title_missing = asin_title_status_df[
    (asin_title_status_df["asin_has_null_title"]) &
    (~asin_title_status_df["asin_has_valid_title"])
]

asin_partial_title = asin_title_status_df[
    (asin_title_status_df["asin_has_null_title"]) &
    (asin_title_status_df["asin_has_valid_title"])
]

In [43]:
# Counts

num_asin_all_title_present = len(asin_all_title_present)
num_asin_all_title_missing = len(asin_all_title_missing)
num_asin_partial_title = len(asin_partial_title)

In [44]:
num_asin_all_title_present

875127

In [45]:
num_asin_all_title_missing

62642

In [46]:
num_asin_partial_title

1313

In [47]:
# Create mapping
title_to_asin_map = (
    amazon_purchases_data
    .dropna(subset=["ASIN/ISBN (Product Code)"])
    .groupby("Title")["ASIN/ISBN (Product Code)"]
    .agg(lambda x: x.mode()[0])   # most frequent ASIN per title
)

In [48]:
# Apply fill (ONLY where ASIN is null)
amazon_purchases_data.loc[
    amazon_purchases_data["ASIN/ISBN (Product Code)"].isna(),
    "ASIN/ISBN (Product Code)"
] = amazon_purchases_data["Title"].map(title_to_asin_map)

In [49]:
# Create mapping for filling missing title
asin_to_title_map = (
    amazon_purchases_data
    .dropna(subset=["Title"])
    .groupby("ASIN/ISBN (Product Code)")["Title"]
    .agg(lambda x: x.mode()[0])
)

In [50]:
# Apply fill
amazon_purchases_data.loc[
    amazon_purchases_data["Title"].isna(),
    "Title"
] = amazon_purchases_data["ASIN/ISBN (Product Code)"].map(asin_to_title_map)

In [51]:
amazon_purchases_data["ASIN/ISBN (Product Code)"].isna().sum()
amazon_purchases_data["Title"].isna().sum()

np.int64(86111)

In [52]:
amazon_purchases_data.isna().sum()

Order Date                      0
Purchase Price Per Unit         0
Quantity                        0
Shipping Address State      24182
Title                       86111
ASIN/ISBN (Product Code)      654
Category                    89458
Survey ResponseID               0
dtype: int64

Alright now time for the product category

In [53]:
# ASIN → Category
# Create mapping
asin_to_category_map = (
    amazon_purchases_data
    .dropna(subset=["Category"])
    .groupby("ASIN/ISBN (Product Code)")["Category"]
    .agg(lambda x: x.mode()[0])
)

In [54]:
# Apply fill
amazon_purchases_data.loc[
    amazon_purchases_data["Category"].isna(),
    "Category"
] = amazon_purchases_data["ASIN/ISBN (Product Code)"].map(asin_to_category_map)

In [55]:
# Title → Category
# Create mapping
title_to_category_map = (
    amazon_purchases_data
    .dropna(subset=["Category"])
    .groupby("Title")["Category"]
    .agg(lambda x: x.mode()[0])
)

In [56]:
# Apply fill
amazon_purchases_data.loc[
    amazon_purchases_data["Category"].isna(),
    "Category"
] = amazon_purchases_data["Title"].map(title_to_category_map)

In [57]:
amazon_purchases_data["Category"].isna().sum()

np.int64(85866)

In [58]:
servey_data.isna().sum()

Survey ResponseID                0
Q-demos-age                      0
Q-demos-hispanic                 0
Q-demos-race                     0
Q-demos-education                0
Q-demos-income                   0
Q-demos-gender                   0
Q-sexual-orientation             0
Q-demos-state                    0
Q-amazon-use-howmany             0
Q-amazon-use-hh-size             0
Q-amazon-use-how-oft             0
Q-substance-use-cigarettes       0
Q-substance-use-marijuana        0
Q-substance-use-alcohol          0
Q-personal-diabetes              0
Q-personal-wheelchair            0
Q-life-changes                3384
Q-sell-YOUR-data                 0
Q-sell-consumer-data             0
Q-small-biz-use                  0
Q-census-use                     0
Q-research-society               0
dtype: int64

It might be safe to assume that the life changes part here with null values basically mean no. Lets confirm that by looking at all the unique items in that section

In [59]:
servey_data["Q-life-changes"].unique()

array(['Lost a job ', nan, 'Moved place of residence', 'Had a child',
       'Lost a job ,Divorce', 'Became pregnant',
       'Moved place of residence,Became pregnant',
       'Lost a job ,Moved place of residence', 'Divorce',
       'Lost a job ,Divorce,Moved place of residence',
       'Became pregnant,Had a child',
       'Moved place of residence,Had a child',
       'Lost a job ,Moved place of residence,Became pregnant',
       'Lost a job ,Became pregnant,Had a child',
       'Divorce,Moved place of residence',
       'Moved place of residence,Became pregnant,Had a child',
       'Lost a job ,Moved place of residence,Had a child',
       'Lost a job ,Had a child', 'Lost a job ,Became pregnant',
       'Lost a job ,Moved place of residence,Became pregnant,Had a child',
       'Divorce,Moved place of residence,Became pregnant',
       'Divorce,Moved place of residence,Became pregnant,Had a child',
       'Lost a job ,Divorce,Moved place of residence,Became pregnant'],
      dtype=

In [60]:
servey_data["Q-life-changes"].nunique()

22

In [61]:
servey_data["Q-life-changes"].value_counts()

Q-life-changes
Moved place of residence                                            808
Lost a job                                                          347
Lost a job ,Moved place of residence                                183
Had a child                                                          63
Became pregnant                                                      57
Became pregnant,Had a child                                          33
Divorce,Moved place of residence                                     22
Divorce                                                              18
Moved place of residence,Had a child                                 16
Moved place of residence,Became pregnant,Had a child                 15
Moved place of residence,Became pregnant                             13
Lost a job ,Had a child                                              12
Lost a job ,Divorce,Moved place of residence                         12
Lost a job ,Divorce                              

In [62]:
servey_data["Q-life-changes"].value_counts(dropna=False)

Q-life-changes
NaN                                                                 3384
Moved place of residence                                             808
Lost a job                                                           347
Lost a job ,Moved place of residence                                 183
Had a child                                                           63
Became pregnant                                                       57
Became pregnant,Had a child                                           33
Divorce,Moved place of residence                                      22
Divorce                                                               18
Moved place of residence,Had a child                                  16
Moved place of residence,Became pregnant,Had a child                  15
Moved place of residence,Became pregnant                              13
Lost a job ,Had a child                                               12
Lost a job ,Divorce,Moved place of r

From this it is clear that Null values basically mean No life changes

In [63]:
servey_data["Q-life-changes"] = servey_data["Q-life-changes"].fillna("No")

In [64]:
servey_data["Q-life-changes"].value_counts(dropna=False)

Q-life-changes
No                                                                  3384
Moved place of residence                                             808
Lost a job                                                           347
Lost a job ,Moved place of residence                                 183
Had a child                                                           63
Became pregnant                                                       57
Became pregnant,Had a child                                           33
Divorce,Moved place of residence                                      22
Divorce                                                               18
Moved place of residence,Had a child                                  16
Moved place of residence,Became pregnant,Had a child                  15
Moved place of residence,Became pregnant                              13
Lost a job ,Had a child                                               12
Lost a job ,Divorce,Moved place of r